### PR TITLE
refactor(nvd): migrate to API 2.0

### PR DESCRIPTION
**Pull request by DmitriyLewen**
## Description
Migrate to NVD API 2.0

## Related Issues
- https://github.com/aquasecurity/trivy/issues/5658

## Related PRs
- aquasecurity/vuln-list-update/pull/258

### CODE REVIEW

**Comment by namandf** [comment]
Hi @DmitriyLewen , @knqyf263 , Hope you are doing great.

Just came across this PR.

1. Do we intend to bring in this change before 15th?
2. In case , we don't get this change by 15th which is the deadline for deprecation and 18th December NVD will stop supporting older feeds if i am not wrong, then what will be impact on trivy scans? Will we continue seeing older vulnerabilities while the new ones or updates will be missing?

Thank you

**Comment by DmitriyLewen** [comment]
Hello @namandf
 
> Do we intend to bring in this change before 15th?

We are trying to finish these changes before the 15th.

> Will this fix require a trivy upgrade? I am assuming no, because its just a change in the way we gather data right?

right. This fix only for trivy-db. You will need only doewnload new DB.

> In case , we don't get this change by 15th which is the deadline for deprecation and 18th December NVD will stop supporting older feeds if i am not wrong, then what will be impact on trivy scans? Will we continue seeing older vulnerabilities while the new ones or updates will be missing?

We only receive advisory information from nvd (severity, descriptions, etc.).
In this case, you will get all old and new CVEs, but there may be some typos/omissions in the vulnerability information.

**Comment by namandf** [comment]
Thank you for the update @DmitriyLewen .

> We only receive advisory information from nvd (severity, descriptions, etc.). In this case, you will get all old and new CVEs, but there may be some typos/omissions in the vulnerability information.
Out of curiosity, do we rely on MITRE/cve.org for the CVE list? or you suggesting that other databases bridge that gap? 
MITRE also seems to have gone through a similar change. https://www.cve.org/Media/News/item/blog/2023/07/25/Legacy-Downloads-being-Phased-Out

**Comment by DmitriyLewen** [comment]
> Out of curiosity, do we rely on MITRE/cve.org for the CVE list? 

No, we don't use MITRE/cve.org.

We use the following databases to get CVE list:
https://aquasecurity.github.io/trivy/v0.48/docs/scanner/vulnerability/#data-sources
https://aquasecurity.github.io/trivy/v0.48/docs/scanner/vulnerability/#data-sources_1

**Comment by namandf** [comment]
> > Out of curiosity, do we rely on MITRE/cve.org for the CVE list?
> 
> No, we don't use MITRE/cve.org.
> 
> We use the following databases to get CVE list: https://aquasecurity.github.io/trivy/v0.48/docs/scanner/vulnerability/#data-sources https://aquasecurity.github.io/trivy/v0.48/docs/scanner/vulnerability/#data-sources_1

Got it. Thank you.

You might already be aware but looks like there is again a change in deadline.
Screenshot_2023_1215_075423

**Comment by DmitriyLewen** [comment]
yes, thanks!

FYI - We have https://github.com/aquasecurity/trivy/issues/5658

File: pkg/vulnsrc/nvd/nvd.go [hunk 1]
```diff
@@ -120,13 +121,48 @@ func (vs VulnSrc) commit(tx *bolt.Tx, items []Item) error {
...
 		return xerrors.Errorf("error in batch update: %w", err)
 	}
 	return nil
 }
+
+// getCvssV2 selects vector, score and severity from V2 metrics
+func getCvssV2(metricsV2 []CvssMetricV2) (score float64, vector string, severity types.Severity) {
+	for _, metricV2 := range metricsV2 {
+		// save first metric or the `Primary` metric if `Secondary` metric was saved previously
+		if score == 0 || metricV2.Type == primaryType {
```
Comment (knqyf263) [inline pkg/vulnsrc/nvd/nvd.go:139]: If I understand correctly, the type doesn't matter. We need metrics from NVD, so I think we should take metrics with `source: nvd@nist.gov`.

>source identifies the organization that provided the metrics information and type identifies whether the organization is a primary or secondary source. Primary sources include the NVD and CNA who have reached the provider level in CVMAP. 10% of provider level submissions are audited by the NVD. If a submission has been audited the NVD will appear as the primary source and the provider level CNA will appear as the secondary source.

```
"cvssMetricV30": [
                        {
                            "source": "secalert@redhat.com",
                            "type": "Secondary",
                            "cvssData": {
                                "version": "3.0",
                                "vectorString": "CVSS:3.0/AV:P/AC:H/PR:N/UI:N/S:U/C:N/I:L/A:H",
                                "attackVector": "PHYSICAL",
                                "attackComplexity": "HIGH",
                                "privilegesRequired": "NONE",
                                "userInteraction": "NONE",
                                "scope": "UNCHANGED",
                                "confidentialityImpact": "NONE",
                                "integrityImpact": "LOW",
                                "availabilityImpact": "HIGH",
                                "baseScore": 4.8,
                                "baseSeverity": "MEDIUM"
                            },
                            "exploitabilityScore": 0.5,
                            "impactScore": 4.2
                        }
                    ],
```

File: pkg/vulnsrc/nvd/nvd.go [hunk 2]
```diff
@@ -120,13 +121,48 @@ func (vs VulnSrc) commit(tx *bolt.Tx, items []Item) error {
...
+			score = metricV3.CvssData.BaseScore
+			vector = metricV3.CvssData.VectorString
+			severity, _ = types.NewSeverity(metricV3.CvssData.BaseSeverity)
+			if metricV3.Type == primaryType {
+				return
+			}
+		}
+	}
+	return
+}
```
[the remark this answers is not in the document]
Comment (knqyf263) [reply pkg/vulnsrc/nvd/nvd.go:168]: As I commented above, we should take metrics only from NVD here. So, the answer is:

1. cvssMetricV31 with `source: nvd@nist.gov`
2. cvssMetricV30 with `source: nvd@nist.gov`
3. cvssMetricV31 from other vendors (ignore)
4. cvssMetricV30 from other vendors (ignore)

We may want to use metrics from other vendors in the future, but it should be done in another PR.

[the remark this answers is not in the document]
Comment (DmitriyLewen) [reply pkg/vulnsrc/nvd/nvd.go]: Thanks for your help!
Changed in https://github.com/aquasecurity/trivy-db/pull/374/commits/96b8f7d8cbc3446b5807f6815edaa0c705dbdb52

File: pkg/vulnsrc/nvd/nvd.go (hunk 1 above)
Comment (DmitriyLewen) [reply pkg/vulnsrc/nvd/nvd.go]: Thanks! fixed in https://github.com/aquasecurity/trivy-db/pull/374/commits/96b8f7d8cbc3446b5807f6815edaa0c705dbdb52

File: pkg/vulnsrc/nvd/types.go [hunk 3]
```diff
@@ -1,79 +1,156 @@
...
+	EvaluatorSolution string          `json:"evaluatorSolution,omitempty"`
+	EvaluatorImpact   string          `json:"evaluatorImpact,omitempty"`
+	CisaExploitAdd    string          `json:"cisaExploitAdd,omitempty"`
+	CisaActionDue     string          `json:"cisaActionDue,omitempty"`
+	Descriptions      []LangString    `json:"descriptions"`
+	Metrics           Metrics         `json:"metrics,omitempty"`
+	Weaknesses        []Weakness      `json:"weaknesses,omitempty"`
+	Configurations    []Configuration `json:"configurations,omitempty"`
+	References        []Reference     `json:"references"`
+	VendorComments    []VendorComment `json:"vendorComments,omitempty"`
```
Comment (knqyf263) [inline pkg/vulnsrc/nvd/types.go:20]: Can we remove unneeded fields? Other structs as well.

**Review by knqyf263 — COMMENTED** [review]
LGTM. I left a small comment.

**Review by knqyf263 — APPROVED** [review]

File: pkg/vulnsrc/nvd/types.go (hunk 3 above)
Comment (knqyf263) [reply pkg/vulnsrc/nvd/types.go]: Removed. Please let me know if I did something wrong.

Comment (DmitriyLewen) [reply pkg/vulnsrc/nvd/types.go]: Good point!
I checked your changes and don't see problem with that.

**Pull request merged**